### PR TITLE
improve configs

### PR DESCRIPTION
**Pull request by tanner0101**
This PR overhauls the provider/configuration system in Vapor to solve the following five problems.

### 1: Minimal dependency injection

Currently, if a user needs access to a property on the Droplet, they need to pass the entire Droplet. This is done to prevent having an old pointer to the property if the one on the Droplet changes.

With this update, all properties on the Droplet are constant meaning you can safely pass any of them to other code.

### 2: Middleware Ordering

Currently it is impossible to respect the ordering of the middleware supplied in the config files due to the dynamic nature of middleware.

With this update, the ordering of middleware from the config is respected.

### 3: Config Dependencies

Currently it is impossible to have a configurable property (such as a logger) depend on another configurable property (such as a console). This is currently being worked around by adding things in the `beforeRun` method. https://github.com/vapor/fluent-provider/blob/master/Sources/FluentProvider/Provider.swift#L155

With this update you can specifically resolve dependencies you need using `config.resolve(...)`. As long as there is not a dependency loop, it will work.

```swift
final class NeedsLoggerMiddleware: Middleware, ConfigInitializable {
    let log: LogProtocol
    
    init(config: Config) throws {
        self.log = try config.resolve(LogProtocol.self)
    }
    
    ...
}
```

### 4: Validate Config

Currently it is impossible to validate whether the Config supplied to a Droplet is valid (i.e., whether the keys specified are spelled correctly.)

With this update, you now get errors like:

```
Configuration error: A Middleware named 'session' (chosen at middleware in Config/droplet.json) was not found (available: error, file, sessions, date).
```

### 5: Less bloat

The config system has built up quite a bit of bloat overtime and is overdue for simplification. This PR is less overall lines of non-test code. Further cleanup is still possible.

Addresses https://github.com/vapor/vapor/issues/971.

### CODE REVIEW

File: Sources/Vapor/Config/Configurable.swift [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        return c
+    }
+    
+    // MARK: Utilities
+    
+    private func key<C>(for type: C.Type, name: String? = nil) throws -> String {
+        let key: String
+        
+        switch type {
+        case is LogProtocol.Type, LogProtocol.self:
```
Comment (LoganWright) [inline Sources/Vapor/Config/Configurable.swift]: what is the comma for here, haven't seent hat before

File: Sources/Vapor/Provider/Config+Provider.swift [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import Configs
+
+private let providersKey = "vapor:providers"
+
+extension Config {
+    var providers: [Provider] {
```
Comment (LoganWright) [inline Sources/Vapor/Provider/Config+Provider.swift]: pub?

File: Sources/Vapor/Provider/Config+Provider.swift [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        get { return storage[providersKey] as? [Provider] ?? [] }
+        set { storage[providersKey] = newValue }
+    }
+    
+    public mutating func addProvider<P: Provider>(_ provider: P) throws {
```
Comment (LoganWright) [inline Sources/Vapor/Provider/Config+Provider.swift]: should these just be `add(_ `?

File: Sources/Vapor/Config/Configurable.swift (hunk 1 above)
Comment (tanner0101) [inline Sources/Vapor/Config/Configurable.swift]: yeah that was a bit messy. i've restructured it so we don't need this switch at all.

File: Sources/Configs/Config.swift [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            guard parts.count == 2 else {
+                return nil
+            }
+            
+            return parts.first
+        }
+        
+        let workDir = arguments.value(for: "workdir")
+            ?? arguments.value(for: "workDir")
+            ?? fileWorkDirectory()
```
Comment (tanner0101) [inline Sources/Configs/Config.swift]: nice! swapped it out

**Pull request merged**